# MSDS692 Project: Car Accidents in Denver (January 2014- December 2018)
## Natalia Weakly 
### Individual accidents
### Merging Data and Preparing for ML

In [1]:
#Imports
import pandas as pd
import numpy as np
import os
import datetime
import seaborn as sns
import matplotlib.pyplot as plt

## Load the Main Car Accident Data
### Add Features

In [2]:
# Load data
accidents_all=pd.read_csv('../MSDSProject/accidents_all.csv', index_col=0, parse_dates=[8, 9])

In [3]:
# copy records beginning January 1, 2014 to a new data frame
accidents_all2 = accidents_all[accidents_all['FIRST_OCCURENCE_DATE'] >= '2014-01-01 00:00:00']
# reset indexes on a new dataframe
accidents_all2=accidents_all2.reset_index(drop=True)

In [4]:
#check load results
accidents_all2.head()

OBJECTID_1  INCIDENT_ID    OFFENSE_ID  OFFENSE_CODE  \
0       41743      2014118  201411854010          5401   
1       41745      2014235  201423554410          5441   
2       41746      2014304  201430454410          5441   
3       41748      2014479  201447954010          5401   
4       41749      2014519  201451954410          5441   

   OFFENSE_CODE_EXTENSION OFFENSE_TYPE_ID          OFFENSE_CATEGORY_ID  \
0                       0  TRAF-HIT & RUN  TRAF - ACCIDENT - HIT & RUN   
1                       0    TRF-ACCIDENT              TRAF - ACCIDENT   
2                       0    TRF-ACCIDENT              TRAF - ACCIDENT   
3                       0  TRAF-HIT & RUN  TRAF - ACCIDENT - HIT & RUN   
4                       0    TRF-ACCIDENT              TRAF - ACCIDENT   

  FIRST_OCCURENCE_DATE       REPORTED_DATE              INCIDENT_ADDRESS  \
0  2014-01-01 00:40:00 2014-01-01 01:05:00                 1520 STOUT ST   
1  2014-01-01 01:35:00 2014-01-01 01:35:00  E COLFAX AVE / N HUMBOLDT ST   
2  2014-01-01 02:04:00 2014-01-01 04:00:00        W COLFAX AVE / 15TH ST   
3  2014-01-01 03:30:00 2014-01-01 04:24:00    W 29TH AVE / N LOWELL BLVD   
4  2014-01-01 03:50:00 2014-01-01 03:58:00    E 53RD AVE / N CHAMBERS RD   

       GEO_X      GEO_Y     GEO_LON    GEO_LAT  DISTRICT_ID  PRECINCT_ID  \
0  3142399.0  1696650.0 -104.993603  39.745019          6.0        611.0   
1  3149134.0  1694869.0 -104.969690  39.740025          6.0        622.0   
2  3143834.0  1695035.0 -104.988533  39.740564          6.0        611.0   
3  3130841.0  1701506.0 -105.034616  39.758519          1.0        113.0   
4  3193971.0  1713504.0 -104.809755  39.790353          5.0        512.0   

  NEIGHBORHOOD_ID  BICYCLE_IND  PEDESTRIAN_IND  
0             CBD            0               0  
1   Cheesman Park            0               0  
2    Civic Center            0               0  
3   West Highland            0               0  
4       Montbello            0               0

In [6]:
accidents_all2.tail()

OBJECTID_1  INCIDENT_ID        OFFENSE_ID  OFFENSE_CODE  \
122480      166997     20198445     2019844554410          5441   
122481      166999     20198720     2019872054410          5441   
122482      167000   2018871461   201887146154010          5401   
122483      167002  20198000502  2019800050254410          5441   
122484      167003  20198000600  2019800060054410          5441   

        OFFENSE_CODE_EXTENSION OFFENSE_TYPE_ID          OFFENSE_CATEGORY_ID  \
122480                       0    TRF-ACCIDENT              TRAF - ACCIDENT   
122481                       0    TRF-ACCIDENT              TRAF - ACCIDENT   
122482                       0  TRAF-HIT & RUN  TRAF - ACCIDENT - HIT & RUN   
122483                       0    TRF-ACCIDENT              TRAF - ACCIDENT   
122484                       0    TRF-ACCIDENT              TRAF - ACCIDENT   

       FIRST_OCCURENCE_DATE       REPORTED_DATE              INCIDENT_ADDRESS  \
122480  2019-01-04 17:27:00 2019-01-04 18:19:00     W 14TH AVE / N RALEIGH ST   
122481  2019-01-04 19:09:00 2019-01-04 19:51:00        I25 HWYNB / W 23RD AVE   
122482  2018-12-24 08:00:00 2018-12-28 13:04:00       3500 BLOCK S TAMARAC DR   
122483  2019-01-04 06:05:00 2019-01-04 06:06:00  6400 BLOCK PENA BLVD INBOUND   
122484  2019-01-04 16:35:00 2019-01-04 17:47:00         6975 N VALLEY HEAD ST   

            GEO_X      GEO_Y     GEO_LON    GEO_LAT  DISTRICT_ID  PRECINCT_ID  \
122480  3128897.0  1694196.0 -105.041662  39.738479          1.0        122.0   
122481  3135940.0  1698910.0 -105.016530  39.751320          1.0        123.0   
122482  3169313.0  1663007.0 -104.898687  39.652212          3.0        324.0   
122483  3200490.0  1724090.0 -104.786262  39.819274          5.0        522.0   
122484  3234312.0  1724993.0 -104.665834  39.820962          7.0        759.0   

       NEIGHBORHOOD_ID  BICYCLE_IND  PEDESTRIAN_IND  
122480     West Colfax            0               0  
122481  Jefferson Park            0               0  
122482   Hampden South            0               0  
122483             DIA            0               0  
122484             DIA            0               0

In [7]:
accidents_all2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 122485 entries, 0 to 122484
Data columns (total 19 columns):
OBJECTID_1                122485 non-null int64
INCIDENT_ID               122485 non-null int64
OFFENSE_ID                122485 non-null int64
OFFENSE_CODE              122485 non-null int64
OFFENSE_CODE_EXTENSION    122485 non-null int64
OFFENSE_TYPE_ID           122485 non-null object
OFFENSE_CATEGORY_ID       122485 non-null object
FIRST_OCCURENCE_DATE      122485 non-null datetime64[ns]
REPORTED_DATE             122485 non-null datetime64[ns]
INCIDENT_ADDRESS          122485 non-null object
GEO_X                     122485 non-null float64
GEO_Y                     122485 non-null float64
GEO_LON                   122485 non-null float64
GEO_LAT                   122485 non-null float64
DISTRICT_ID               122406 non-null float64
PRECINCT_ID               122406 non-null float64
NEIGHBORHOOD_ID           122485 non-null object
BICYCLE_IND               122485 non-nu

In [8]:
#Drop 'INCIDENT_ADDRESS', 'REPORTED_DATE', 'DISTRICT_ID', 'PRECINCT_ID'
#Drop 'GEO_X', 'GEO_Y' - duplicate information in a different coordinate system
accidents_all2.drop('INCIDENT_ID', axis=1, inplace=True)
accidents_all2.drop('OFFENSE_ID', axis=1, inplace=True)
accidents_all2.drop('INCIDENT_ADDRESS', axis=1, inplace=True)
accidents_all2.drop('REPORTED_DATE', axis=1, inplace=True)
accidents_all2.drop('GEO_X', axis=1, inplace=True)
accidents_all2.drop('GEO_Y', axis=1, inplace=True)
accidents_all2.drop('DISTRICT_ID', axis=1, inplace=True)
accidents_all2.drop('PRECINCT_ID', axis=1, inplace=True)

In [9]:
#Check results
accidents_all2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 122485 entries, 0 to 122484
Data columns (total 11 columns):
OBJECTID_1                122485 non-null int64
OFFENSE_CODE              122485 non-null int64
OFFENSE_CODE_EXTENSION    122485 non-null int64
OFFENSE_TYPE_ID           122485 non-null object
OFFENSE_CATEGORY_ID       122485 non-null object
FIRST_OCCURENCE_DATE      122485 non-null datetime64[ns]
GEO_LON                   122485 non-null float64
GEO_LAT                   122485 non-null float64
NEIGHBORHOOD_ID           122485 non-null object
BICYCLE_IND               122485 non-null int64
PEDESTRIAN_IND            122485 non-null int64
dtypes: datetime64[ns](1), float64(2), int64(5), object(3)
memory usage: 10.3+ MB


In [10]:
#Extract accident date (drop time)
accidents_all2['Date'] = accidents_all2['FIRST_OCCURENCE_DATE'].dt.date

In [14]:
#convert extracted date to datetime format
accidents_all2['Date'] = pd.to_datetime(accidents_all2['Date'])

In [15]:
#add dayOfWeek column
#Monday=0, Sunday =6
accidents_all2['dayOfWeek']= accidents_all2['FIRST_OCCURENCE_DATE'].dt.dayofweek

In [16]:
#Convert 'dayOfWeek' to category
accidents_all2['dayOfWeek'] = accidents_all2['dayOfWeek'].astype('category')

In [17]:
accidents_all2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 122485 entries, 0 to 122484
Data columns (total 13 columns):
OBJECTID_1                122485 non-null int64
OFFENSE_CODE              122485 non-null int64
OFFENSE_CODE_EXTENSION    122485 non-null int64
OFFENSE_TYPE_ID           122485 non-null object
OFFENSE_CATEGORY_ID       122485 non-null object
FIRST_OCCURENCE_DATE      122485 non-null datetime64[ns]
GEO_LON                   122485 non-null float64
GEO_LAT                   122485 non-null float64
NEIGHBORHOOD_ID           122485 non-null object
BICYCLE_IND               122485 non-null int64
PEDESTRIAN_IND            122485 non-null int64
Date                      122485 non-null datetime64[ns]
dayOfWeek                 122485 non-null category
dtypes: category(1), datetime64[ns](2), float64(2), int64(5), object(3)
memory usage: 11.3+ MB


In [18]:
#Convert 'OFFENSE_TYPE_ID' to categories
accidents_all2['OFFENSE_TYPE_ID']=accidents_all2['OFFENSE_TYPE_ID'].astype('category') 

In [19]:
#Check results
accidents_all2['OFFENSE_TYPE_ID']

0         TRAF-HIT & RUN
1           TRF-ACCIDENT
2           TRF-ACCIDENT
3         TRAF-HIT & RUN
4           TRF-ACCIDENT
5           TRF-ACCIDENT
6           TRF-ACCIDENT
7           TRF-ACCIDENT
8           TRF-ACCIDENT
9           TRF-ACCIDENT
10          TRF-ACCIDENT
11        TRAF-HIT & RUN
12          TRF-ACCIDENT
13          TRF-ACCIDENT
14          TRF-ACCIDENT
15        TRAF-HIT & RUN
16          TRF-ACCIDENT
17        TRAF-HIT & RUN
18          TRF-ACCIDENT
19          TRF-ACCIDENT
20        TRAF-HIT & RUN
21          TRF-ACCIDENT
22          TRF-ACCIDENT
23          TRF-ACCIDENT
24          TRF-ACCIDENT
25        TRAF-HIT & RUN
26        TRAF-HIT & RUN
27          TRF-ACCIDENT
28          TRF-ACCIDENT
29          TRF-ACCIDENT
               ...      
122455      TRF-ACCIDENT
122456      TRF-ACCIDENT
122457      TRF-ACCIDENT
122458      TRF-ACCIDENT
122459    TRAF-HIT & RUN
122460      TRF-ACCIDENT
122461    TRAF-HIT & RUN
122462      TRF-ACCIDENT
122463      TRF-ACCIDENT


In [20]:
#Convert 'OFFENSE_CATEGORY_ID' to categories
accidents_all2['OFFENSE_CATEGORY_ID']=accidents_all2['OFFENSE_CATEGORY_ID'].astype('category') 

In [21]:
#Check results
accidents_all2['OFFENSE_CATEGORY_ID']

0         TRAF - ACCIDENT - HIT & RUN
1                     TRAF - ACCIDENT
2                     TRAF - ACCIDENT
3         TRAF - ACCIDENT - HIT & RUN
4                     TRAF - ACCIDENT
5                     TRAF - ACCIDENT
6                     TRAF - ACCIDENT
7                     TRAF - ACCIDENT
8                     TRAF - ACCIDENT
9                     TRAF - ACCIDENT
10                    TRAF - ACCIDENT
11        TRAF - ACCIDENT - HIT & RUN
12                    TRAF - ACCIDENT
13                    TRAF - ACCIDENT
14                    TRAF - ACCIDENT
15        TRAF - ACCIDENT - HIT & RUN
16                    TRAF - ACCIDENT
17        TRAF - ACCIDENT - HIT & RUN
18                    TRAF - ACCIDENT
19                    TRAF - ACCIDENT
20        TRAF - ACCIDENT - HIT & RUN
21                    TRAF - ACCIDENT
22                    TRAF - ACCIDENT
23                    TRAF - ACCIDENT
24                    TRAF - ACCIDENT
25        TRAF - ACCIDENT - HIT & RUN
26        TR

In [22]:
#Convert 'OFFENSE_CODE' to categories
accidents_all2['OFFENSE_CODE']=accidents_all2['OFFENSE_CODE'].astype('category') 

In [23]:
#Check results
accidents_all2['OFFENSE_CODE']

0         5401
1         5441
2         5441
3         5401
4         5441
5         5441
6         5441
7         5441
8         5441
9         5441
10        5441
11        5401
12        5441
13        5441
14        5441
15        5401
16        5441
17        5401
18        5441
19        5441
20        5401
21        5441
22        5441
23        5441
24        5441
25        5401
26        5401
27        5441
28        5441
29        5441
          ... 
122455    5441
122456    5441
122457    5441
122458    5441
122459    5401
122460    5441
122461    5401
122462    5441
122463    5441
122464    5441
122465    5401
122466    5441
122467    5401
122468    5441
122469    5401
122470    5401
122471    5401
122472    5401
122473    5441
122474    5441
122475    5401
122476    5441
122477    5441
122478    5401
122479    5441
122480    5441
122481    5441
122482    5401
122483    5441
122484    5441
Name: OFFENSE_CODE, Length: 122485, dtype: category
Categories (3, int64): [5401, 542


NCIC (National Crime Information Center Codes):
- 5401 - Hit and Run
- 5420 - Accident Fatal
- 5441 - Accident Damage

Source: http://www.aurorafop49.org/2015%20NCIC%20sheet%20%202.pdf


In [24]:
#Convert 'OFFENSE_CODE_EXTENSION' to categories
accidents_all2['OFFENSE_CODE_EXTENSION']=accidents_all2['OFFENSE_CODE_EXTENSION'].astype('category')

In [25]:
#Check results
accidents_all2['OFFENSE_CODE_EXTENSION']

0         0
1         0
2         0
3         0
4         0
5         0
6         0
7         0
8         0
9         0
10        0
11        0
12        0
13        0
14        0
15        0
16        0
17        0
18        0
19        0
20        0
21        0
22        0
23        0
24        0
25        0
26        0
27        0
28        0
29        0
         ..
122455    0
122456    0
122457    0
122458    0
122459    0
122460    0
122461    0
122462    0
122463    0
122464    0
122465    0
122466    0
122467    0
122468    0
122469    0
122470    0
122471    0
122472    0
122473    0
122474    0
122475    0
122476    0
122477    0
122478    0
122479    0
122480    0
122481    0
122482    0
122483    0
122484    0
Name: OFFENSE_CODE_EXTENSION, Length: 122485, dtype: category
Categories (4, int64): [0, 1, 2, 3]

In [26]:
#Convert 'NEIGHBORHOOD_ID' to categories 
accidents_all2['NEIGHBORHOOD_ID']=accidents_all2['NEIGHBORHOOD_ID'].astype('category')

In [27]:
#Check results 
accidents_all2['NEIGHBORHOOD_ID']

0                                  CBD
1                        Cheesman Park
2                         Civic Center
3                        West Highland
4                            Montbello
5                  Northeast Park Hill
6                         Lincoln Park
7                          East Colfax
8                     Virginia Village
9                          West Colfax
10                            Valverde
11                         West Colfax
12                                 CBD
13                          University
14                    Virginia Village
15                             Hampden
16                        Lincoln Park
17                       West Highland
18                             Windsor
19                         West Colfax
20                               Speer
21                        Lincoln Park
22                         Barnum West
23        Gateway - Green Valley Ranch
24                           Stapleton
25                       

In [28]:
#Create a column with values: 1 - at least one bicylce was involved, 0 - no bicycles involved
accidents_all2['BicycleInvolved']=accidents_all2['BICYCLE_IND'].apply(lambda x: 1 if x > 0 else 0 )

In [29]:
#Number of accidents involving bicycles
accidents_all2['BicycleInvolved'].sum()

1667

In [30]:
#Create a column with values: 1 at least one pedestrian was involved, 0 - no pedestrians involved
accidents_all2['PedestrianInvolved']=accidents_all2['PEDESTRIAN_IND'].apply(lambda x: 1 if x > 0 else 0 )

In [31]:
accidents_all2['PedestrianInvolved'].sum()

2766

In [32]:
# Delete original columns describing the involvement of bicyclists and pedestrians
accidents_all2.drop('PEDESTRIAN_IND', axis=1, inplace=True)
accidents_all2.drop('BICYCLE_IND', axis=1, inplace=True)

In [33]:
# Check results
accidents_all2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 122485 entries, 0 to 122484
Data columns (total 13 columns):
OBJECTID_1                122485 non-null int64
OFFENSE_CODE              122485 non-null category
OFFENSE_CODE_EXTENSION    122485 non-null category
OFFENSE_TYPE_ID           122485 non-null category
OFFENSE_CATEGORY_ID       122485 non-null category
FIRST_OCCURENCE_DATE      122485 non-null datetime64[ns]
GEO_LON                   122485 non-null float64
GEO_LAT                   122485 non-null float64
NEIGHBORHOOD_ID           122485 non-null category
Date                      122485 non-null datetime64[ns]
dayOfWeek                 122485 non-null category
BicycleInvolved           122485 non-null int64
PedestrianInvolved        122485 non-null int64
dtypes: category(6), datetime64[ns](2), float64(2), int64(3)
memory usage: 7.2 MB


In [34]:
accidents_all2.head()

OBJECTID_1 OFFENSE_CODE OFFENSE_CODE_EXTENSION OFFENSE_TYPE_ID  \
0       41743         5401                      0  TRAF-HIT & RUN   
1       41745         5441                      0    TRF-ACCIDENT   
2       41746         5441                      0    TRF-ACCIDENT   
3       41748         5401                      0  TRAF-HIT & RUN   
4       41749         5441                      0    TRF-ACCIDENT   

           OFFENSE_CATEGORY_ID FIRST_OCCURENCE_DATE     GEO_LON    GEO_LAT  \
0  TRAF - ACCIDENT - HIT & RUN  2014-01-01 00:40:00 -104.993603  39.745019   
1              TRAF - ACCIDENT  2014-01-01 01:35:00 -104.969690  39.740025   
2              TRAF - ACCIDENT  2014-01-01 02:04:00 -104.988533  39.740564   
3  TRAF - ACCIDENT - HIT & RUN  2014-01-01 03:30:00 -105.034616  39.758519   
4              TRAF - ACCIDENT  2014-01-01 03:50:00 -104.809755  39.790353   

  NEIGHBORHOOD_ID       Date dayOfWeek  BicycleInvolved  PedestrianInvolved  
0             CBD 2014-01-01         2                0                   0  
1   Cheesman Park 2014-01-01         2                0                   0  
2    Civic Center 2014-01-01         2                0                   0  
3   West Highland 2014-01-01         2                0                   0  
4       Montbello 2014-01-01         2                0                   0

In [35]:
accidents_all2.tail()

OBJECTID_1 OFFENSE_CODE OFFENSE_CODE_EXTENSION OFFENSE_TYPE_ID  \
122480      166997         5441                      0    TRF-ACCIDENT   
122481      166999         5441                      0    TRF-ACCIDENT   
122482      167000         5401                      0  TRAF-HIT & RUN   
122483      167002         5441                      0    TRF-ACCIDENT   
122484      167003         5441                      0    TRF-ACCIDENT   

                OFFENSE_CATEGORY_ID FIRST_OCCURENCE_DATE     GEO_LON  \
122480              TRAF - ACCIDENT  2019-01-04 17:27:00 -105.041662   
122481              TRAF - ACCIDENT  2019-01-04 19:09:00 -105.016530   
122482  TRAF - ACCIDENT - HIT & RUN  2018-12-24 08:00:00 -104.898687   
122483              TRAF - ACCIDENT  2019-01-04 06:05:00 -104.786262   
122484              TRAF - ACCIDENT  2019-01-04 16:35:00 -104.665834   

          GEO_LAT NEIGHBORHOOD_ID       Date dayOfWeek  BicycleInvolved  \
122480  39.738479     West Colfax 2019-01-04         4                0   
122481  39.751320  Jefferson Park 2019-01-04         4                0   
122482  39.652212   Hampden South 2018-12-24         0                0   
122483  39.819274             DIA 2019-01-04         4                0   
122484  39.820962             DIA 2019-01-04         4                0   

        PedestrianInvolved  
122480                   0  
122481                   0  
122482                   0  
122483                   0  
122484                   0

### Load Holidays 

In [36]:
#Load csv file
holidays = pd.read_csv('../MSDSProject/holidays.csv',index_col=0, parse_dates=[1])

In [37]:
#check load results
holidays.head()

Date         Federal_Holiday
0 2014-01-01          New Year's Day
1 2014-01-20  Martin Luther King Day
2 2014-02-17         Presidents' Day
3 2014-05-26            Memorial Day
4 2014-07-04        Independence Day

In [38]:
holidays.tail()

Date   Federal_Holiday
60 2019-09-02         Labor Day
61 2019-10-14      Columbus Day
62 2019-11-11      Veterans Day
63 2019-11-28  Thanksgiving Day
64 2019-12-25     Christmas Day

In [39]:
holidays.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 65 entries, 0 to 64
Data columns (total 2 columns):
Date               65 non-null datetime64[ns]
Federal_Holiday    65 non-null object
dtypes: datetime64[ns](1), object(1)
memory usage: 1.5+ KB


In [40]:
# Merge accidents data and the holidays dataframe 
accidents_all3 = accidents_all2.merge(holidays, how='left', on='Date')

In [41]:
# Check results
accidents_all3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 122485 entries, 0 to 122484
Data columns (total 14 columns):
OBJECTID_1                122485 non-null int64
OFFENSE_CODE              122485 non-null category
OFFENSE_CODE_EXTENSION    122485 non-null category
OFFENSE_TYPE_ID           122485 non-null category
OFFENSE_CATEGORY_ID       122485 non-null category
FIRST_OCCURENCE_DATE      122485 non-null datetime64[ns]
GEO_LON                   122485 non-null float64
GEO_LAT                   122485 non-null float64
NEIGHBORHOOD_ID           122485 non-null category
Date                      122485 non-null datetime64[ns]
dayOfWeek                 122485 non-null category
BicycleInvolved           122485 non-null int64
PedestrianInvolved        122485 non-null int64
Federal_Holiday           2911 non-null object
dtypes: category(6), datetime64[ns](2), float64(2), int64(3), object(1)
memory usage: 9.1+ MB


In [42]:
accidents_all3.head()

OBJECTID_1 OFFENSE_CODE OFFENSE_CODE_EXTENSION OFFENSE_TYPE_ID  \
0       41743         5401                      0  TRAF-HIT & RUN   
1       41745         5441                      0    TRF-ACCIDENT   
2       41746         5441                      0    TRF-ACCIDENT   
3       41748         5401                      0  TRAF-HIT & RUN   
4       41749         5441                      0    TRF-ACCIDENT   

           OFFENSE_CATEGORY_ID FIRST_OCCURENCE_DATE     GEO_LON    GEO_LAT  \
0  TRAF - ACCIDENT - HIT & RUN  2014-01-01 00:40:00 -104.993603  39.745019   
1              TRAF - ACCIDENT  2014-01-01 01:35:00 -104.969690  39.740025   
2              TRAF - ACCIDENT  2014-01-01 02:04:00 -104.988533  39.740564   
3  TRAF - ACCIDENT - HIT & RUN  2014-01-01 03:30:00 -105.034616  39.758519   
4              TRAF - ACCIDENT  2014-01-01 03:50:00 -104.809755  39.790353   

  NEIGHBORHOOD_ID       Date dayOfWeek  BicycleInvolved  PedestrianInvolved  \
0             CBD 2014-01-01         2                0                   0   
1   Cheesman Park 2014-01-01         2                0                   0   
2    Civic Center 2014-01-01         2                0                   0   
3   West Highland 2014-01-01         2                0                   0   
4       Montbello 2014-01-01         2                0                   0   

  Federal_Holiday  
0  New Year's Day  
1  New Year's Day  
2  New Year's Day  
3  New Year's Day  
4  New Year's Day

In [43]:
accidents_all3.tail()

OBJECTID_1 OFFENSE_CODE OFFENSE_CODE_EXTENSION OFFENSE_TYPE_ID  \
122480      166997         5441                      0    TRF-ACCIDENT   
122481      166999         5441                      0    TRF-ACCIDENT   
122482      167000         5401                      0  TRAF-HIT & RUN   
122483      167002         5441                      0    TRF-ACCIDENT   
122484      167003         5441                      0    TRF-ACCIDENT   

                OFFENSE_CATEGORY_ID FIRST_OCCURENCE_DATE     GEO_LON  \
122480              TRAF - ACCIDENT  2019-01-04 17:27:00 -105.041662   
122481              TRAF - ACCIDENT  2019-01-04 19:09:00 -105.016530   
122482  TRAF - ACCIDENT - HIT & RUN  2018-12-24 08:00:00 -104.898687   
122483              TRAF - ACCIDENT  2019-01-04 06:05:00 -104.786262   
122484              TRAF - ACCIDENT  2019-01-04 16:35:00 -104.665834   

          GEO_LAT NEIGHBORHOOD_ID       Date dayOfWeek  BicycleInvolved  \
122480  39.738479     West Colfax 2019-01-04         4                0   
122481  39.751320  Jefferson Park 2019-01-04         4                0   
122482  39.652212   Hampden South 2018-12-24         0                0   
122483  39.819274             DIA 2019-01-04         4                0   
122484  39.820962             DIA 2019-01-04         4                0   

        PedestrianInvolved Federal_Holiday  
122480                   0             NaN  
122481                   0             NaN  
122482                   0             NaN  
122483                   0             NaN  
122484                   0             NaN

In [44]:
accidents_all3['Federal_Holiday'].fillna('not holiday', inplace=True)

In [45]:
#Create the "Holiday' column with values: 1  = official holiday, 0 - not holiday
accidents_all3['Holiday'] = accidents_all3['Federal_Holiday'].map(lambda x: 0 if x == 'not holiday' in x else 1)

In [46]:
#Check results
accidents_all3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 122485 entries, 0 to 122484
Data columns (total 15 columns):
OBJECTID_1                122485 non-null int64
OFFENSE_CODE              122485 non-null category
OFFENSE_CODE_EXTENSION    122485 non-null category
OFFENSE_TYPE_ID           122485 non-null category
OFFENSE_CATEGORY_ID       122485 non-null category
FIRST_OCCURENCE_DATE      122485 non-null datetime64[ns]
GEO_LON                   122485 non-null float64
GEO_LAT                   122485 non-null float64
NEIGHBORHOOD_ID           122485 non-null category
Date                      122485 non-null datetime64[ns]
dayOfWeek                 122485 non-null category
BicycleInvolved           122485 non-null int64
PedestrianInvolved        122485 non-null int64
Federal_Holiday           122485 non-null object
Holiday                   122485 non-null int64
dtypes: category(6), datetime64[ns](2), float64(2), int64(4), object(1)
memory usage: 10.0+ MB


In [47]:
#look at the data
accidents_all3.head()

OBJECTID_1 OFFENSE_CODE OFFENSE_CODE_EXTENSION OFFENSE_TYPE_ID  \
0       41743         5401                      0  TRAF-HIT & RUN   
1       41745         5441                      0    TRF-ACCIDENT   
2       41746         5441                      0    TRF-ACCIDENT   
3       41748         5401                      0  TRAF-HIT & RUN   
4       41749         5441                      0    TRF-ACCIDENT   

           OFFENSE_CATEGORY_ID FIRST_OCCURENCE_DATE     GEO_LON    GEO_LAT  \
0  TRAF - ACCIDENT - HIT & RUN  2014-01-01 00:40:00 -104.993603  39.745019   
1              TRAF - ACCIDENT  2014-01-01 01:35:00 -104.969690  39.740025   
2              TRAF - ACCIDENT  2014-01-01 02:04:00 -104.988533  39.740564   
3  TRAF - ACCIDENT - HIT & RUN  2014-01-01 03:30:00 -105.034616  39.758519   
4              TRAF - ACCIDENT  2014-01-01 03:50:00 -104.809755  39.790353   

  NEIGHBORHOOD_ID       Date dayOfWeek  BicycleInvolved  PedestrianInvolved  \
0             CBD 2014-01-01         2                0                   0   
1   Cheesman Park 2014-01-01         2                0                   0   
2    Civic Center 2014-01-01         2                0                   0   
3   West Highland 2014-01-01         2                0                   0   
4       Montbello 2014-01-01         2                0                   0   

  Federal_Holiday  Holiday  
0  New Year's Day        1  
1  New Year's Day        1  
2  New Year's Day        1  
3  New Year's Day        1  
4  New Year's Day        1

In [48]:
#drop the "Federal_Holiday" column
accidents_all3.drop(['Federal_Holiday'], axis=1, inplace=True)

In [49]:
#Check results
accidents_all3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 122485 entries, 0 to 122484
Data columns (total 14 columns):
OBJECTID_1                122485 non-null int64
OFFENSE_CODE              122485 non-null category
OFFENSE_CODE_EXTENSION    122485 non-null category
OFFENSE_TYPE_ID           122485 non-null category
OFFENSE_CATEGORY_ID       122485 non-null category
FIRST_OCCURENCE_DATE      122485 non-null datetime64[ns]
GEO_LON                   122485 non-null float64
GEO_LAT                   122485 non-null float64
NEIGHBORHOOD_ID           122485 non-null category
Date                      122485 non-null datetime64[ns]
dayOfWeek                 122485 non-null category
BicycleInvolved           122485 non-null int64
PedestrianInvolved        122485 non-null int64
Holiday                   122485 non-null int64
dtypes: category(6), datetime64[ns](2), float64(2), int64(4)
memory usage: 9.1 MB


### Rush Hour Traffic

In [50]:
# extract time of the accident
accidents_all3['Time'] = accidents_all3['FIRST_OCCURENCE_DATE'].dt.time

In [51]:
# check results
accidents_all3.head()

OBJECTID_1 OFFENSE_CODE OFFENSE_CODE_EXTENSION OFFENSE_TYPE_ID  \
0       41743         5401                      0  TRAF-HIT & RUN   
1       41745         5441                      0    TRF-ACCIDENT   
2       41746         5441                      0    TRF-ACCIDENT   
3       41748         5401                      0  TRAF-HIT & RUN   
4       41749         5441                      0    TRF-ACCIDENT   

           OFFENSE_CATEGORY_ID FIRST_OCCURENCE_DATE     GEO_LON    GEO_LAT  \
0  TRAF - ACCIDENT - HIT & RUN  2014-01-01 00:40:00 -104.993603  39.745019   
1              TRAF - ACCIDENT  2014-01-01 01:35:00 -104.969690  39.740025   
2              TRAF - ACCIDENT  2014-01-01 02:04:00 -104.988533  39.740564   
3  TRAF - ACCIDENT - HIT & RUN  2014-01-01 03:30:00 -105.034616  39.758519   
4              TRAF - ACCIDENT  2014-01-01 03:50:00 -104.809755  39.790353   

  NEIGHBORHOOD_ID       Date dayOfWeek  BicycleInvolved  PedestrianInvolved  \
0             CBD 2014-01-01         2                0                   0   
1   Cheesman Park 2014-01-01         2                0                   0   
2    Civic Center 2014-01-01         2                0                   0   
3   West Highland 2014-01-01         2                0                   0   
4       Montbello 2014-01-01         2                0                   0   

   Holiday      Time  
0        1  00:40:00  
1        1  01:35:00  
2        1  02:04:00  
3        1  03:30:00  
4        1  03:50:00

In [52]:
# Function to check if accident time falls within traditional rush hours  07:00:00 - 09:00:00 and 16:00:00 - 18:00:00
# Note: this function checks only times and does not discriminate between weekdays, weekends and holidays

def rush_hour(x):
    """ checks if timestamp falls within rush hour periods (07:00:00 - 09:00:00 and 16:00:00 - 18:00:00)
    """
    start1 = datetime.time(7, 0,0)
    end1 = datetime.time(9,0,0)
    start2 = datetime.time(16,0,0)
    end2 = datetime.time(18,0,0)
    if x >= start1:
        if x <= end1:
            return 1
        elif x >= start2:
            if x <= end2:
                return 1       
    return 0

In [53]:
# Create the "RushHour' column with values: 1  = accident happened during rush hours, 0 - not during rush hours
# Let's consider times from 07:00:00 to 09:00:00 and from 16:00:00 to 18:00:00 On 24-hour clock to be rush hour traffic 
accidents_all3['RushHour'] = accidents_all3['Time'].apply(lambda x: rush_hour(x))
# Drop the 'Time' column
accidents_all3.drop('Time', axis=1, inplace=True)

In [54]:
# Check results
accidents_all3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 122485 entries, 0 to 122484
Data columns (total 15 columns):
OBJECTID_1                122485 non-null int64
OFFENSE_CODE              122485 non-null category
OFFENSE_CODE_EXTENSION    122485 non-null category
OFFENSE_TYPE_ID           122485 non-null category
OFFENSE_CATEGORY_ID       122485 non-null category
FIRST_OCCURENCE_DATE      122485 non-null datetime64[ns]
GEO_LON                   122485 non-null float64
GEO_LAT                   122485 non-null float64
NEIGHBORHOOD_ID           122485 non-null category
Date                      122485 non-null datetime64[ns]
dayOfWeek                 122485 non-null category
BicycleInvolved           122485 non-null int64
PedestrianInvolved        122485 non-null int64
Holiday                   122485 non-null int64
RushHour                  122485 non-null int64
dtypes: category(6), datetime64[ns](2), float64(2), int64(5)
memory usage: 10.0 MB


In [55]:
# Check results
accidents_all3.head()

OBJECTID_1 OFFENSE_CODE OFFENSE_CODE_EXTENSION OFFENSE_TYPE_ID  \
0       41743         5401                      0  TRAF-HIT & RUN   
1       41745         5441                      0    TRF-ACCIDENT   
2       41746         5441                      0    TRF-ACCIDENT   
3       41748         5401                      0  TRAF-HIT & RUN   
4       41749         5441                      0    TRF-ACCIDENT   

           OFFENSE_CATEGORY_ID FIRST_OCCURENCE_DATE     GEO_LON    GEO_LAT  \
0  TRAF - ACCIDENT - HIT & RUN  2014-01-01 00:40:00 -104.993603  39.745019   
1              TRAF - ACCIDENT  2014-01-01 01:35:00 -104.969690  39.740025   
2              TRAF - ACCIDENT  2014-01-01 02:04:00 -104.988533  39.740564   
3  TRAF - ACCIDENT - HIT & RUN  2014-01-01 03:30:00 -105.034616  39.758519   
4              TRAF - ACCIDENT  2014-01-01 03:50:00 -104.809755  39.790353   

  NEIGHBORHOOD_ID       Date dayOfWeek  BicycleInvolved  PedestrianInvolved  \
0             CBD 2014-01-01         2                0                   0   
1   Cheesman Park 2014-01-01         2                0                   0   
2    Civic Center 2014-01-01         2                0                   0   
3   West Highland 2014-01-01         2                0                   0   
4       Montbello 2014-01-01         2                0                   0   

   Holiday  RushHour  
0        1         0  
1        1         0  
2        1         0  
3        1         0  
4        1         0

In [56]:
accidents_all3.tail()

OBJECTID_1 OFFENSE_CODE OFFENSE_CODE_EXTENSION OFFENSE_TYPE_ID  \
122480      166997         5441                      0    TRF-ACCIDENT   
122481      166999         5441                      0    TRF-ACCIDENT   
122482      167000         5401                      0  TRAF-HIT & RUN   
122483      167002         5441                      0    TRF-ACCIDENT   
122484      167003         5441                      0    TRF-ACCIDENT   

                OFFENSE_CATEGORY_ID FIRST_OCCURENCE_DATE     GEO_LON  \
122480              TRAF - ACCIDENT  2019-01-04 17:27:00 -105.041662   
122481              TRAF - ACCIDENT  2019-01-04 19:09:00 -105.016530   
122482  TRAF - ACCIDENT - HIT & RUN  2018-12-24 08:00:00 -104.898687   
122483              TRAF - ACCIDENT  2019-01-04 06:05:00 -104.786262   
122484              TRAF - ACCIDENT  2019-01-04 16:35:00 -104.665834   

          GEO_LAT NEIGHBORHOOD_ID       Date dayOfWeek  BicycleInvolved  \
122480  39.738479     West Colfax 2019-01-04         4                0   
122481  39.751320  Jefferson Park 2019-01-04         4                0   
122482  39.652212   Hampden South 2018-12-24         0                0   
122483  39.819274             DIA 2019-01-04         4                0   
122484  39.820962             DIA 2019-01-04         4                0   

        PedestrianInvolved  Holiday  RushHour  
122480                   0        0         1  
122481                   0        0         0  
122482                   0        0         1  
122483                   0        0         0  
122484                   0        0         1

### Load the Natural Light Conditions Data

In [57]:
# Load the csv file
light=pd.read_csv('../MSDSProject/naturalLightConditions.csv',index_col=0, parse_dates=[1, 2, 3, 4, 5, 6, 7, 8, 9])

In [58]:
# Check load results
light.head()

fullDate        AstroT_Start          AstroT_End     NauticalT_Start  \
0 2014-01-01 2014-01-01 05:44:00 2014-01-01 18:23:00 2014-01-01 06:17:00   
1 2014-01-02 2014-01-02 05:44:00 2014-01-02 18:23:00 2014-01-02 06:17:00   
2 2014-01-03 2014-01-03 05:44:00 2014-01-03 18:24:00 2014-01-03 06:17:00   
3 2014-01-04 2014-01-04 05:44:00 2014-01-04 18:25:00 2014-01-04 06:17:00   
4 2014-01-05 2014-01-05 05:44:00 2014-01-05 18:26:00 2014-01-05 06:17:00   

        NauticalT_End        CivilT_Start          CivilT_End  \
0 2014-01-01 17:50:00 2014-01-01 06:50:00 2014-01-01 17:16:00   
1 2014-01-02 17:51:00 2014-01-02 06:50:00 2014-01-02 17:17:00   
2 2014-01-03 17:51:00 2014-01-03 06:51:00 2014-01-03 17:18:00   
3 2014-01-04 17:52:00 2014-01-04 06:51:00 2014-01-04 17:19:00   
4 2014-01-05 17:53:00 2014-01-05 06:51:00 2014-01-05 17:19:00   

              Sunrise              Sunset  
0 2014-01-01 07:21:00 2014-01-01 16:46:00  
1 2014-01-02 07:21:00 2014-01-02 16:47:00  
2 2014-01-03 07:21:00 2014-01-03 16:48:00  
3 2014-01-04 07:21:00 2014-01-04 16:48:00  
4 2014-01-05 07:21:00 2014-01-05 16:49:00

In [59]:
light.tail()

fullDate        AstroT_Start          AstroT_End     NauticalT_Start  \
1911 2019-03-27 2019-03-27 05:21:00 2019-03-27 20:50:00 2019-03-27 05:53:00   
1912 2019-03-28 2019-03-28 05:19:00 2019-03-28 20:51:00 2019-03-28 05:51:00   
1913 2019-03-29 2019-03-29 05:17:00 2019-03-29 20:52:00 2019-03-29 05:50:00   
1914 2019-03-30 2019-03-30 05:15:00 2019-03-30 20:54:00 2019-03-30 05:48:00   
1915 2019-03-31 2019-03-31 05:13:00 2019-03-31 20:55:00 2019-03-31 05:46:00   

           NauticalT_End        CivilT_Start          CivilT_End  \
1911 2019-03-27 20:17:00 2019-03-27 06:25:00 2019-03-27 19:46:00   
1912 2019-03-28 20:18:00 2019-03-28 06:23:00 2019-03-28 19:47:00   
1913 2019-03-29 20:20:00 2019-03-29 06:22:00 2019-03-29 19:48:00   
1914 2019-03-30 20:21:00 2019-03-30 06:20:00 2019-03-30 19:49:00   
1915 2019-03-31 20:22:00 2019-03-31 06:18:00 2019-03-31 19:50:00   

                 Sunrise              Sunset  
1911 2019-03-27 06:52:00 2019-03-27 19:18:00  
1912 2019-03-28 06:50:00 2019-03-28 19:19:00  
1913 2019-03-29 06:49:00 2019-03-29 19:20:00  
1914 2019-03-30 06:47:00 2019-03-30 19:21:00  
1915 2019-03-31 06:45:00 2019-03-31 19:22:00

In [61]:
light.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1916 entries, 0 to 1915
Data columns (total 9 columns):
fullDate           1916 non-null datetime64[ns]
AstroT_Start       1916 non-null datetime64[ns]
AstroT_End         1916 non-null datetime64[ns]
NauticalT_Start    1916 non-null datetime64[ns]
NauticalT_End      1916 non-null datetime64[ns]
CivilT_Start       1916 non-null datetime64[ns]
CivilT_End         1916 non-null datetime64[ns]
Sunrise            1916 non-null datetime64[ns]
Sunset             1916 non-null datetime64[ns]
dtypes: datetime64[ns](9)
memory usage: 149.7 KB


In [62]:
# merge accidents data and information about natural light conditions 
accidents_all4 = pd.merge(accidents_all3, light, how='left', left_on='Date', right_on='fullDate')
# drop 'fullDate' column
accidents_all4.drop(['fullDate'], axis=1, inplace=True)

In [63]:
#check results
accidents_all4.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 122485 entries, 0 to 122484
Data columns (total 23 columns):
OBJECTID_1                122485 non-null int64
OFFENSE_CODE              122485 non-null category
OFFENSE_CODE_EXTENSION    122485 non-null category
OFFENSE_TYPE_ID           122485 non-null category
OFFENSE_CATEGORY_ID       122485 non-null category
FIRST_OCCURENCE_DATE      122485 non-null datetime64[ns]
GEO_LON                   122485 non-null float64
GEO_LAT                   122485 non-null float64
NEIGHBORHOOD_ID           122485 non-null category
Date                      122485 non-null datetime64[ns]
dayOfWeek                 122485 non-null category
BicycleInvolved           122485 non-null int64
PedestrianInvolved        122485 non-null int64
Holiday                   122485 non-null int64
RushHour                  122485 non-null int64
AstroT_Start              122485 non-null datetime64[ns]
AstroT_End                122485 non-null datetime64[ns]
NauticalT_Start

In [64]:
accidents_all4.head()

OBJECTID_1 OFFENSE_CODE OFFENSE_CODE_EXTENSION OFFENSE_TYPE_ID  \
0       41743         5401                      0  TRAF-HIT & RUN   
1       41745         5441                      0    TRF-ACCIDENT   
2       41746         5441                      0    TRF-ACCIDENT   
3       41748         5401                      0  TRAF-HIT & RUN   
4       41749         5441                      0    TRF-ACCIDENT   

           OFFENSE_CATEGORY_ID FIRST_OCCURENCE_DATE     GEO_LON    GEO_LAT  \
0  TRAF - ACCIDENT - HIT & RUN  2014-01-01 00:40:00 -104.993603  39.745019   
1              TRAF - ACCIDENT  2014-01-01 01:35:00 -104.969690  39.740025   
2              TRAF - ACCIDENT  2014-01-01 02:04:00 -104.988533  39.740564   
3  TRAF - ACCIDENT - HIT & RUN  2014-01-01 03:30:00 -105.034616  39.758519   
4              TRAF - ACCIDENT  2014-01-01 03:50:00 -104.809755  39.790353   

  NEIGHBORHOOD_ID       Date         ...         Holiday  RushHour  \
0             CBD 2014-01-01         ...               1         0   
1   Cheesman Park 2014-01-01         ...               1         0   
2    Civic Center 2014-01-01         ...               1         0   
3   West Highland 2014-01-01         ...               1         0   
4       Montbello 2014-01-01         ...               1         0   

         AstroT_Start          AstroT_End     NauticalT_Start  \
0 2014-01-01 05:44:00 2014-01-01 18:23:00 2014-01-01 06:17:00   
1 2014-01-01 05:44:00 2014-01-01 18:23:00 2014-01-01 06:17:00   
2 2014-01-01 05:44:00 2014-01-01 18:23:00 2014-01-01 06:17:00   
3 2014-01-01 05:44:00 2014-01-01 18:23:00 2014-01-01 06:17:00   
4 2014-01-01 05:44:00 2014-01-01 18:23:00 2014-01-01 06:17:00   

        NauticalT_End        CivilT_Start          CivilT_End  \
0 2014-01-01 17:50:00 2014-01-01 06:50:00 2014-01-01 17:16:00   
1 2014-01-01 17:50:00 2014-01-01 06:50:00 2014-01-01 17:16:00   
2 2014-01-01 17:50:00 2014-01-01 06:50:00 2014-01-01 17:16:00   
3 2014-01-01 17:50:00 2014-01-01 06:50:00 2014-01-01 17:16:00   
4 2014-01-01 17:50:00 2014-01-01 06:50:00 2014-01-01 17:16:00   

              Sunrise              Sunset  
0 2014-01-01 07:21:00 2014-01-01 16:46:00  
1 2014-01-01 07:21:00 2014-01-01 16:46:00  
2 2014-01-01 07:21:00 2014-01-01 16:46:00  
3 2014-01-01 07:21:00 2014-01-01 16:46:00  
4 2014-01-01 07:21:00 2014-01-01 16:46:00  

[5 rows x 23 columns]

In [65]:
accidents_all4.tail()

OBJECTID_1 OFFENSE_CODE OFFENSE_CODE_EXTENSION OFFENSE_TYPE_ID  \
122480      166997         5441                      0    TRF-ACCIDENT   
122481      166999         5441                      0    TRF-ACCIDENT   
122482      167000         5401                      0  TRAF-HIT & RUN   
122483      167002         5441                      0    TRF-ACCIDENT   
122484      167003         5441                      0    TRF-ACCIDENT   

                OFFENSE_CATEGORY_ID FIRST_OCCURENCE_DATE     GEO_LON  \
122480              TRAF - ACCIDENT  2019-01-04 17:27:00 -105.041662   
122481              TRAF - ACCIDENT  2019-01-04 19:09:00 -105.016530   
122482  TRAF - ACCIDENT - HIT & RUN  2018-12-24 08:00:00 -104.898687   
122483              TRAF - ACCIDENT  2019-01-04 06:05:00 -104.786262   
122484              TRAF - ACCIDENT  2019-01-04 16:35:00 -104.665834   

          GEO_LAT NEIGHBORHOOD_ID       Date         ...         Holiday  \
122480  39.738479     West Colfax 2019-01-04         ...               0   
122481  39.751320  Jefferson Park 2019-01-04         ...               0   
122482  39.652212   Hampden South 2018-12-24         ...               0   
122483  39.819274             DIA 2019-01-04         ...               0   
122484  39.820962             DIA 2019-01-04         ...               0   

        RushHour        AstroT_Start          AstroT_End     NauticalT_Start  \
122480         1 2019-01-04 05:44:00 2019-01-04 18:25:00 2019-01-04 06:17:00   
122481         0 2019-01-04 05:44:00 2019-01-04 18:25:00 2019-01-04 06:17:00   
122482         1 2018-12-24 05:41:00 2018-12-24 18:17:00 2018-12-24 06:14:00   
122483         0 2019-01-04 05:44:00 2019-01-04 18:25:00 2019-01-04 06:17:00   
122484         1 2019-01-04 05:44:00 2019-01-04 18:25:00 2019-01-04 06:17:00   

             NauticalT_End        CivilT_Start          CivilT_End  \
122480 2019-01-04 17:52:00 2019-01-04 06:51:00 2019-01-04 17:18:00   
122481 2019-01-04 17:52:00 2019-01-04 06:51:00 2019-01-04 17:18:00   
122482 2018-12-24 17:44:00 2018-12-24 06:48:00 2018-12-24 17:10:00   
122483 2019-01-04 17:52:00 2019-01-04 06:51:00 2019-01-04 17:18:00   
122484 2019-01-04 17:52:00 2019-01-04 06:51:00 2019-01-04 17:18:00   

                   Sunrise              Sunset  
122480 2019-01-04 07:21:00 2019-01-04 16:48:00  
122481 2019-01-04 07:21:00 2019-01-04 16:48:00  
122482 2018-12-24 07:18:00 2018-12-24 16:40:00  
122483 2019-01-04 07:21:00 2019-01-04 16:48:00  
122484 2019-01-04 07:21:00 2019-01-04 16:48:00  

[5 rows x 23 columns]



Use four periods in the day depending on natural light conditions:

- astronomical twilight (astroT) = [from astroT Start to mauticalT start] AND [from nauticalT end to astroT end]
- nautical twilight (nauticalT) = [from nauticalT start to civilT start] AND [from CivilT end to nauticalT end]
- civil twilight= [from civil twilight start to sunrise] AND [from sunset to civil twilight end] 
- daytime (day) = [from sunrise to sunset]

If an observation does not have '1' in either of twilight periods or daytime period columns, it means that this accident happened during the night time conditions: night = [from to 00:00:00 to Astronomical twilight start] AND [ after astronomical twilight end to 23:59:59]



In [66]:
# Functions to check if accident happened during daylight or any of the special 
# twilight periods

def daylight(x):
    """ returns 1 if timestamp falls within daytime
    """
    if x['FIRST_OCCURENCE_DATE'] >= x['Sunrise']:
        if x['FIRST_OCCURENCE_DATE'] < x['Sunset']:
            return 1
    return 0
    
    
def astro_twilight(x):
    """ checks if timestamp falls within astronomical twighlight periods
    """
    if x['FIRST_OCCURENCE_DATE'] >= x['AstroT_Start']:
        if x['FIRST_OCCURENCE_DATE'] < x['NauticalT_Start']:
            return 1
        elif x['FIRST_OCCURENCE_DATE'] >= x['NauticalT_End']:
            if x['FIRST_OCCURENCE_DATE'] < x['AstroT_End']:
                return 1       
    return 0

def nautical_twilight(x):
    """ checks if timestamp falls within nautical twighlight periods
    """
    if x['FIRST_OCCURENCE_DATE'] >= x['NauticalT_Start']:
        if x['FIRST_OCCURENCE_DATE'] < x['CivilT_Start']:
            return 1
        elif x['FIRST_OCCURENCE_DATE'] >= x['CivilT_End']:
            if x['FIRST_OCCURENCE_DATE'] < x['NauticalT_End']:
                return 1       
    return 0

def civil_twilight(x):
    """ checks if timestamp falls within civil twighlight periods
    """
    if x['FIRST_OCCURENCE_DATE'] >= x['CivilT_Start']:
        if x['FIRST_OCCURENCE_DATE'] < x['Sunrise']:
            return 1
        elif x['FIRST_OCCURENCE_DATE'] >= x['Sunset']:
            if x['FIRST_OCCURENCE_DATE'] < x['CivilT_End']:
                return 1       
    return 0

In [67]:
# Create the "Daytime' column with values: 1  = accident happened during daytime, 0 - not during daytime 
accidents_all4['Daytime'] = accidents_all4.apply(daylight, axis=1)

In [68]:
# Create the "AstroTwilight' column with values: 1  = accident happened during astronimical twilight time, 0 - not during astronomical twitlight time 
accidents_all4['AstroTwilight'] = accidents_all4.apply(astro_twilight, axis=1)

In [69]:
# Create the "NauticalTwilight' column with values: 1  = accident happened during nautical twilight time, 0 - not during nautical twitlight time 
accidents_all4['NauticalTwilight'] = accidents_all4.apply(nautical_twilight, axis=1)

In [70]:
# Create the "CivilTwilight' column with values: 1  = accident happened during civil twilight time, 0 - not during civil twitlight time 
accidents_all4['CivilTwilight'] = accidents_all4.apply(civil_twilight, axis=1)

In [71]:
# Drop columns containing initial information about natural light conditions
natlight_cols=['AstroT_Start', 'AstroT_End', 'NauticalT_Start', 'NauticalT_End', 'CivilT_Start', 'CivilT_End', 'Sunrise', 'Sunset']
accidents_all4.drop(natlight_cols, axis=1, inplace=True)

In [72]:
# Check results
accidents_all4.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 122485 entries, 0 to 122484
Data columns (total 19 columns):
OBJECTID_1                122485 non-null int64
OFFENSE_CODE              122485 non-null category
OFFENSE_CODE_EXTENSION    122485 non-null category
OFFENSE_TYPE_ID           122485 non-null category
OFFENSE_CATEGORY_ID       122485 non-null category
FIRST_OCCURENCE_DATE      122485 non-null datetime64[ns]
GEO_LON                   122485 non-null float64
GEO_LAT                   122485 non-null float64
NEIGHBORHOOD_ID           122485 non-null category
Date                      122485 non-null datetime64[ns]
dayOfWeek                 122485 non-null category
BicycleInvolved           122485 non-null int64
PedestrianInvolved        122485 non-null int64
Holiday                   122485 non-null int64
RushHour                  122485 non-null int64
Daytime                   122485 non-null int64
AstroTwilight             122485 non-null int64
NauticalTwilight          122485 

In [73]:
accidents_all4.head()

OBJECTID_1 OFFENSE_CODE OFFENSE_CODE_EXTENSION OFFENSE_TYPE_ID  \
0       41743         5401                      0  TRAF-HIT & RUN   
1       41745         5441                      0    TRF-ACCIDENT   
2       41746         5441                      0    TRF-ACCIDENT   
3       41748         5401                      0  TRAF-HIT & RUN   
4       41749         5441                      0    TRF-ACCIDENT   

           OFFENSE_CATEGORY_ID FIRST_OCCURENCE_DATE     GEO_LON    GEO_LAT  \
0  TRAF - ACCIDENT - HIT & RUN  2014-01-01 00:40:00 -104.993603  39.745019   
1              TRAF - ACCIDENT  2014-01-01 01:35:00 -104.969690  39.740025   
2              TRAF - ACCIDENT  2014-01-01 02:04:00 -104.988533  39.740564   
3  TRAF - ACCIDENT - HIT & RUN  2014-01-01 03:30:00 -105.034616  39.758519   
4              TRAF - ACCIDENT  2014-01-01 03:50:00 -104.809755  39.790353   

  NEIGHBORHOOD_ID       Date dayOfWeek  BicycleInvolved  PedestrianInvolved  \
0             CBD 2014-01-01         2                0                   0   
1   Cheesman Park 2014-01-01         2                0                   0   
2    Civic Center 2014-01-01         2                0                   0   
3   West Highland 2014-01-01         2                0                   0   
4       Montbello 2014-01-01         2                0                   0   

   Holiday  RushHour  Daytime  AstroTwilight  NauticalTwilight  CivilTwilight  
0        1         0        0              0                 0              0  
1        1         0        0              0                 0              0  
2        1         0        0              0                 0              0  
3        1         0        0              0                 0              0  
4        1         0        0              0                 0              0

In [74]:
accidents_all4.tail()

OBJECTID_1 OFFENSE_CODE OFFENSE_CODE_EXTENSION OFFENSE_TYPE_ID  \
122480      166997         5441                      0    TRF-ACCIDENT   
122481      166999         5441                      0    TRF-ACCIDENT   
122482      167000         5401                      0  TRAF-HIT & RUN   
122483      167002         5441                      0    TRF-ACCIDENT   
122484      167003         5441                      0    TRF-ACCIDENT   

                OFFENSE_CATEGORY_ID FIRST_OCCURENCE_DATE     GEO_LON  \
122480              TRAF - ACCIDENT  2019-01-04 17:27:00 -105.041662   
122481              TRAF - ACCIDENT  2019-01-04 19:09:00 -105.016530   
122482  TRAF - ACCIDENT - HIT & RUN  2018-12-24 08:00:00 -104.898687   
122483              TRAF - ACCIDENT  2019-01-04 06:05:00 -104.786262   
122484              TRAF - ACCIDENT  2019-01-04 16:35:00 -104.665834   

          GEO_LAT NEIGHBORHOOD_ID       Date dayOfWeek  BicycleInvolved  \
122480  39.738479     West Colfax 2019-01-04         4                0   
122481  39.751320  Jefferson Park 2019-01-04         4                0   
122482  39.652212   Hampden South 2018-12-24         0                0   
122483  39.819274             DIA 2019-01-04         4                0   
122484  39.820962             DIA 2019-01-04         4                0   

        PedestrianInvolved  Holiday  RushHour  Daytime  AstroTwilight  \
122480                   0        0         1        0              0   
122481                   0        0         0        0              0   
122482                   0        0         1        1              0   
122483                   0        0         0        0              1   
122484                   0        0         1        1              0   

        NauticalTwilight  CivilTwilight  
122480                 1              0  
122481                 0              0  
122482                 0              0  
122483                 0              0  
122484                 0              0

### Load Weather information

Due to the lack of data about the exact weather conditions when an accident occurred, we are going to use daily weather observations for all accidents on that date.


In [75]:
#Load csv file
weather_data=pd.read_csv('../MSDSProject/weatherdata.csv',index_col=0, parse_dates=[1])

In [76]:
#Check load results- dataframe structure
weather_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1853 entries, 0 to 1852
Data columns (total 13 columns):
date                 1853 non-null datetime64[ns]
DIA_max_temp         1853 non-null int64
DIA_min_temp         1853 non-null int64
DIA_precipitation    1853 non-null float64
DIA_snowfall         1853 non-null float64
St_max_temp          1853 non-null float64
St_min_temp          1853 non-null float64
St_precipitation     1853 non-null float64
St_snowfall          1853 non-null float64
DW_max_temp          1853 non-null float64
DW_min_temp          1853 non-null float64
DW_precipitation     1853 non-null float64
DW_snowfall          1853 non-null float64
dtypes: datetime64[ns](1), float64(10), int64(2)
memory usage: 202.7 KB


In [77]:
#Check load results - data
weather_data.head()

date  DIA_max_temp  DIA_min_temp  DIA_precipitation  DIA_snowfall  \
0 2014-01-01            41            13               0.18           0.0   
1 2014-01-02            42            11               0.00           0.0   
2 2014-01-03            62            25               0.00           0.0   
3 2014-01-04            38             6               0.21           0.0   
4 2014-01-05            12           -10               0.05           0.0   

   St_max_temp  St_min_temp  St_precipitation  St_snowfall  DW_max_temp  \
0         57.0         29.0            0.0000       0.0000         62.0   
1         37.0         19.0            0.0400       0.4000         44.0   
2         46.0         16.0            0.0000       0.0000         51.0   
3         63.0         22.0            0.0001       0.0001         64.0   
4         22.0          6.0            0.1500       2.2000         40.0   

   DW_min_temp  DW_precipitation  DW_snowfall  
0         22.0              0.00          0.0  
1         15.0              0.06          0.3  
2         16.0              0.00          0.0  
3         27.0              0.08          1.2  
4         10.0              0.10          2.2

In [78]:
weather_data.tail()

date  DIA_max_temp  DIA_min_temp  DIA_precipitation  DIA_snowfall  \
1848 2019-01-23            41            16               0.00           0.0   
1849 2019-01-24            29            12               0.11           0.0   
1850 2019-01-25            32             8               0.00           0.0   
1851 2019-01-26            42            19               0.00           0.0   
1852 2019-01-27            50            28               0.00           0.0   

      St_max_temp  St_min_temp  St_precipitation  St_snowfall  DW_max_temp  \
1848         31.0         16.0              0.01          0.1         33.0   
1849         46.0         20.0              0.00          0.0         49.0   
1850         29.0         11.0              0.07          1.5         33.0   
1851         38.0         11.0              0.00          0.0         44.0   
1852         45.0         22.0              0.00          0.0         49.0   

      DW_min_temp  DW_precipitation  DW_snowfall  
1848         17.0            0.0001       0.0001  
1849         24.0            0.0000       0.0000  
1850         13.0            0.0500       0.5000  
1851         14.0            0.0000       0.0000  
1852         22.0            0.0000       0.0000

In [79]:
# merge accident data and weather data
accidents_all5 = accidents_all4.merge(weather_data, left_on='Date', right_on='date', how='left')
# Drop the duplicate 'date' column
accidents_all5.drop('date', axis=1, inplace=True)

In [81]:
# Drop records for acccidents that occured after the end of 2018
accidents_all6 = accidents_all5[accidents_all5['FIRST_OCCURENCE_DATE'] <= '2018-12-31 23:59:59']
# Reset indexes on the new dataframe
accidents_all6=accidents_all6.reset_index(drop=True)

In [82]:
#Check merge results
accidents_all6.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120056 entries, 0 to 120055
Data columns (total 31 columns):
OBJECTID_1                120056 non-null int64
OFFENSE_CODE              120056 non-null category
OFFENSE_CODE_EXTENSION    120056 non-null category
OFFENSE_TYPE_ID           120056 non-null category
OFFENSE_CATEGORY_ID       120056 non-null category
FIRST_OCCURENCE_DATE      120056 non-null datetime64[ns]
GEO_LON                   120056 non-null float64
GEO_LAT                   120056 non-null float64
NEIGHBORHOOD_ID           120056 non-null category
Date                      120056 non-null datetime64[ns]
dayOfWeek                 120056 non-null category
BicycleInvolved           120056 non-null int64
PedestrianInvolved        120056 non-null int64
Holiday                   120056 non-null int64
RushHour                  120056 non-null int64
Daytime                   120056 non-null int64
AstroTwilight             120056 non-null int64
NauticalTwilight          120056 

In [83]:
accidents_all6.head()

OBJECTID_1 OFFENSE_CODE OFFENSE_CODE_EXTENSION OFFENSE_TYPE_ID  \
0       41743         5401                      0  TRAF-HIT & RUN   
1       41745         5441                      0    TRF-ACCIDENT   
2       41746         5441                      0    TRF-ACCIDENT   
3       41748         5401                      0  TRAF-HIT & RUN   
4       41749         5441                      0    TRF-ACCIDENT   

           OFFENSE_CATEGORY_ID FIRST_OCCURENCE_DATE     GEO_LON    GEO_LAT  \
0  TRAF - ACCIDENT - HIT & RUN  2014-01-01 00:40:00 -104.993603  39.745019   
1              TRAF - ACCIDENT  2014-01-01 01:35:00 -104.969690  39.740025   
2              TRAF - ACCIDENT  2014-01-01 02:04:00 -104.988533  39.740564   
3  TRAF - ACCIDENT - HIT & RUN  2014-01-01 03:30:00 -105.034616  39.758519   
4              TRAF - ACCIDENT  2014-01-01 03:50:00 -104.809755  39.790353   

  NEIGHBORHOOD_ID       Date     ...      DIA_precipitation  DIA_snowfall  \
0             CBD 2014-01-01     ...                   0.18           0.0   
1   Cheesman Park 2014-01-01     ...                   0.18           0.0   
2    Civic Center 2014-01-01     ...                   0.18           0.0   
3   West Highland 2014-01-01     ...                   0.18           0.0   
4       Montbello 2014-01-01     ...                   0.18           0.0   

   St_max_temp  St_min_temp  St_precipitation  St_snowfall  DW_max_temp  \
0         57.0         29.0               0.0          0.0         62.0   
1         57.0         29.0               0.0          0.0         62.0   
2         57.0         29.0               0.0          0.0         62.0   
3         57.0         29.0               0.0          0.0         62.0   
4         57.0         29.0               0.0          0.0         62.0   

   DW_min_temp  DW_precipitation  DW_snowfall  
0         22.0               0.0          0.0  
1         22.0               0.0          0.0  
2         22.0               0.0          0.0  
3         22.0               0.0          0.0  
4         22.0               0.0          0.0  

[5 rows x 31 columns]

In [84]:
accidents_all6.tail()

OBJECTID_1 OFFENSE_CODE OFFENSE_CODE_EXTENSION OFFENSE_TYPE_ID  \
120051      166816         5441                      0    TRF-ACCIDENT   
120052      166817         5401                      0  TRAF-HIT & RUN   
120053      166929         5441                      3  TRAF-ACCIDENT-   
120054      167001         5401                      0  TRAF-HIT & RUN   
120055      167000         5401                      0  TRAF-HIT & RUN   

                OFFENSE_CATEGORY_ID FIRST_OCCURENCE_DATE     GEO_LON  \
120051              TRAF - ACCIDENT  2018-12-31 11:50:00 -104.963127   
120052  TRAF - ACCIDENT - HIT & RUN  2018-12-31 23:28:00 -104.960163   
120053     TRAF - ACCIDENT - POLICE  2018-12-22 20:15:00 -104.977531   
120054  TRAF - ACCIDENT - HIT & RUN  2018-12-30 02:09:00 -104.903460   
120055  TRAF - ACCIDENT - HIT & RUN  2018-12-24 08:00:00 -104.898687   

          GEO_LAT  NEIGHBORHOOD_ID       Date     ...      DIA_precipitation  \
120051  39.739432    Cheesman Park 2018-12-31     ...                 0.0300   
120052  39.727114     Country Club 2018-12-31     ...                 0.0300   
120053  39.738657     Capitol Hill 2018-12-22     ...                 0.0001   
120054  39.758408  North Park Hill 2018-12-30     ...                 0.0000   
120055  39.652212    Hampden South 2018-12-24     ...                 0.0000   

        DIA_snowfall  St_max_temp  St_min_temp  St_precipitation  St_snowfall  \
120051           0.0         53.0         22.0               0.0          0.0   
120052           0.0         53.0         22.0               0.0          0.0   
120053           0.0         62.0         26.0               0.0          0.0   
120054           0.0         41.0          9.0               0.0          0.0   
120055           0.0         52.0         17.0               0.0          0.0   

        DW_max_temp  DW_min_temp  DW_precipitation  DW_snowfall  
120051         55.0         21.0            0.0001       0.0001  
120052         55.0         21.0            0.0001       0.0001  
120053         65.0         28.0            0.0000       0.0000  
120054         42.0          9.0            0.0000       0.0000  
120055         54.0         17.0            0.0000       0.0000  

[5 rows x 31 columns]

### Finishing Touches: Scaling and One Hot Encoding for Categorical Variables

In [85]:
# Scaling of the weather data columns

from sklearn import preprocessing
#Create a MinMax preprocessing object
min_max_scaler = preprocessing.MinMaxScaler()
#columns that need to be scaled
col_toscale=['DIA_max_temp', 'DIA_min_temp', 'DIA_precipitation', 'DIA_snowfall', 'St_max_temp', 'St_min_temp', 'St_precipitation', 'St_snowfall', 'DW_max_temp', 'DW_min_temp', 'DW_precipitation', 'DW_snowfall']

accidents_all6[col_toscale]=min_max_scaler.fit_transform(accidents_all6[col_toscale])

In [86]:
#look at the dataframe
accidents_all6.head()

OBJECTID_1 OFFENSE_CODE OFFENSE_CODE_EXTENSION OFFENSE_TYPE_ID  \
0       41743         5401                      0  TRAF-HIT & RUN   
1       41745         5441                      0    TRF-ACCIDENT   
2       41746         5441                      0    TRF-ACCIDENT   
3       41748         5401                      0  TRAF-HIT & RUN   
4       41749         5441                      0    TRF-ACCIDENT   

           OFFENSE_CATEGORY_ID FIRST_OCCURENCE_DATE     GEO_LON    GEO_LAT  \
0  TRAF - ACCIDENT - HIT & RUN  2014-01-01 00:40:00 -104.993603  39.745019   
1              TRAF - ACCIDENT  2014-01-01 01:35:00 -104.969690  39.740025   
2              TRAF - ACCIDENT  2014-01-01 02:04:00 -104.988533  39.740564   
3  TRAF - ACCIDENT - HIT & RUN  2014-01-01 03:30:00 -105.034616  39.758519   
4              TRAF - ACCIDENT  2014-01-01 03:50:00 -104.809755  39.790353   

  NEIGHBORHOOD_ID       Date     ...      DIA_precipitation  DIA_snowfall  \
0             CBD 2014-01-01     ...                    0.1           0.0   
1   Cheesman Park 2014-01-01     ...                    0.1           0.0   
2    Civic Center 2014-01-01     ...                    0.1           0.0   
3   West Highland 2014-01-01     ...                    0.1           0.0   
4       Montbello 2014-01-01     ...                    0.1           0.0   

   St_max_temp  St_min_temp  St_precipitation  St_snowfall  DW_max_temp  \
0     0.548077      0.52381               0.0          0.0     0.568627   
1     0.548077      0.52381               0.0          0.0     0.568627   
2     0.548077      0.52381               0.0          0.0     0.568627   
3     0.548077      0.52381               0.0          0.0     0.568627   
4     0.548077      0.52381               0.0          0.0     0.568627   

   DW_min_temp  DW_precipitation  DW_snowfall  
0     0.425287               0.0          0.0  
1     0.425287               0.0          0.0  
2     0.425287               0.0          0.0  
3     0.425287               0.0          0.0  
4     0.425287               0.0          0.0  

[5 rows x 31 columns]

In [87]:
# One hot encoding for categorical columns

# Copy columns that need encoding to a separate df
to_encode=accidents_all6[['OFFENSE_CODE', 'OFFENSE_CODE_EXTENSION', 'OFFENSE_TYPE_ID', 'OFFENSE_CATEGORY_ID', 'NEIGHBORHOOD_ID', 'dayOfWeek']]
to_encode.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120056 entries, 0 to 120055
Data columns (total 6 columns):
OFFENSE_CODE              120056 non-null category
OFFENSE_CODE_EXTENSION    120056 non-null category
OFFENSE_TYPE_ID           120056 non-null category
OFFENSE_CATEGORY_ID       120056 non-null category
NEIGHBORHOOD_ID           120056 non-null category
dayOfWeek                 120056 non-null category
dtypes: category(6)
memory usage: 707.7 KB


In [88]:
#one hot encoding
df_dummies =pd.get_dummies(to_encode, prefix=['OffenseCode', 'OffenseExtension', 'OffenseType', 'OffenseCategory', 'Neighboorhood', 'DayOfWeek'])

In [89]:
#Resulting dataframe with dummy variables
df_dummies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120056 entries, 0 to 120055
Columns: 103 entries, OffenseCode_5401 to DayOfWeek_6
dtypes: uint8(103)
memory usage: 11.8 MB


In [90]:
#Add new columns with dummy variables to the main dataframe
accidents_all6=pd.concat([accidents_all6, df_dummies], axis=1)

In [91]:
accidents_all6.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120056 entries, 0 to 120055
Columns: 134 entries, OBJECTID_1 to DayOfWeek_6
dtypes: category(6), datetime64[ns](2), float64(14), int64(9), uint8(103)
memory usage: 35.4 MB


In [92]:
# Drop columns that were used to encode  dummy variables
accidents_all6.drop(['OFFENSE_CODE', 'OFFENSE_CODE_EXTENSION', 'OFFENSE_TYPE_ID', 'OFFENSE_CATEGORY_ID', 'NEIGHBORHOOD_ID', 'dayOfWeek'], axis=1, inplace=True)

In [93]:
#check results
accidents_all6.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120056 entries, 0 to 120055
Columns: 128 entries, OBJECTID_1 to DayOfWeek_6
dtypes: datetime64[ns](2), float64(14), int64(9), uint8(103)
memory usage: 34.7 MB


In [94]:
# Output full accidents dataframe to a file for future use
accidents_all6.to_csv('accidents_all_withIDs.csv', date_format='%Y-%m-%d %H:%M:%S')

In [95]:
# To prepare for clustering delete columns uniquely identifying each incident
accidents_all6.drop(['OBJECTID_1', 'FIRST_OCCURENCE_DATE', 'GEO_LON', 'GEO_LAT', 'Date'], axis=1, inplace=True)
#check results
accidents_all6.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120056 entries, 0 to 120055
Columns: 123 entries, BicycleInvolved to DayOfWeek_6
dtypes: float64(12), int64(8), uint8(103)
memory usage: 30.1 MB


In [96]:
# Output accidents dataframe ready for ML to a file for future use
accidents_all6.to_csv('accidents_allForML.csv', date_format='%Y-%m-%d %H:%M:%S')